### PR TITLE
Abstract out ParameterBuilder and implement it

### LINKED ISSUE #83831: Make ParameterBuilder abstract
With [Abstract System.Reflection.Emit public types](https://github.com/dotnet/runtime/issues/78542) we have abstracted other `System.Reflection.Emit.XyzBuilder` types, `ParameterBuilder` type left out from this effort (probably because it lives in different ref file).

The type also needs to be abstracted same as others because:
 - Because of internal constructor could not create an instance of it outside of CoreLib. (The new `AsemblyBuilder` implementation will be added outside of CoreLib)
 - Even if we workaround the internal constructor we could not get `CustomAttributes` info set to the `ParameterBuilder` instance, same for the default value.

### Suggested API change:
```diff
namespace System.Reflection.Emit
{
-   public class ParameterBuilder
+   public abstract class ParameterBuilder
    {
-       internal ParameterBuilder() { }
+       protected ParameterBuilder() { }
        public virtual int Attributes { get { throw null; } }
        public bool IsIn { get { throw null; } }
        public bool IsOptional { get { throw null; } }
        public bool IsOut { get { throw null; } }
        public virtual string? Name { get { throw null; } }
        public virtual int Position { get { throw null; } }
        public virtual void SetConstant(object? defaultValue) { }
        public void SetCustomAttribute(ConstructorInfo con, byte[] binaryAttribute) { }
+       protected abstract void SetCustomAttributeCore(ConstructorInfo con, byte[] binaryAttribute);
        public void SetCustomAttribute(CustomAttributeBuilder customBuilder) { }
    }
}
```

### PR DESCRIPTION
**Pull request by buyaa-n**
Contributes to https://github.com/dotnet/runtime/issues/62956
Fixes https://github.com/dotnet/runtime/issues/83831
- Abstract out ParameterBuilder and update runtimes accordingly
- Create ParameterBuilderImpl and save methods parameters

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/FieldBuilderImpl.cs [hunk 1]
```diff
@@ -101,386 +101,5 @@ public override void SetValue(object? obj, object? val, BindingFlags invokeAttr,
 
         public override bool IsDefined(Type attributeType, bool inherit) => throw new NotSupportedException(SR.NotSupported_DynamicModule);
         #endregion
-
-        internal sealed class MarshallingInfo
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/FieldBuilderImpl.cs:105]: Moved to `CustomAttributeWrapper.cs` as it is needed for `ParameterBuilderImpl` too

**Comment by vargaz** [comment]
The mono changes look ok.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ParameterBuilderImpl.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+namespace System.Reflection.Emit
+{
+    internal sealed class ParameterBuilderImpl : ParameterBuilder
+    {
+        private readonly string? _name;
+        private readonly int _position;
+        private readonly MethodBuilderImpl _methodBuilder;
+        private ParameterAttributes _attributes;
+
+        internal List<CustomAttributeWrapper>? _customAttributes = new();
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ParameterBuilderImpl.cs:15]:
```suggestion
        internal List<CustomAttributeWrapper>? _customAttributes;
```
?

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs [hunk 3]
```diff
@@ -51,7 +54,17 @@ internal sealed class MethodBuilderImpl : MethodBuilder
 
         protected override bool InitLocalsCore { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
         protected override GenericTypeParameterBuilder[] DefineGenericParametersCore(params string[] names) => throw new NotImplementedException();
-        protected override ParameterBuilder DefineParameterCore(int position, ParameterAttributes attributes, string? strParamName) => throw new NotImplementedException();
+        protected override ParameterBuilder DefineParameterCore(int position, ParameterAttributes attributes, string? strParamName)
+        {
+            if (position > 0 && (_parameterTypes == null || position > _parameterTypes.Length))
+                throw new ArgumentOutOfRangeException(SR.ArgumentOutOfRange_ParamSequence);
+
+            Debug.Assert(_parameters != null);
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs:62]: Is this guaranteed? I think you can end up with `_parameters = null` when position argument is 0

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs [hunk 4]
```diff
@@ -171,7 +173,398 @@ elementType switch
                 PrimitiveSerializationTypeCode.Single => typeof(float),
                 PrimitiveSerializationTypeCode.Double => typeof(double),
                 PrimitiveSerializationTypeCode.String => typeof(string),
-                _ => throw new ArgumentException(SR.Argument_InvalidTypeArgument, "binaryAttribute"),
+                _ => throw new ArgumentException(SR.Argument_InvalidTypeCodeForTypeArgument, "binaryAttribute"),
             };
     }
+
+    internal sealed class MarshallingInfo
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs:180]: Should this be in a separate .cs file?

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs [hunk 5]
```diff
@@ -171,7 +173,398 @@ elementType switch
...
+        private const UnmanagedType InvalidUnmanagedType = (UnmanagedType)Invalid;
+        private const VarEnum InvalidVariantType = (VarEnum)Invalid;
+        private const int MaxMarshalInteger = 0x1fffffff;
+
+        internal BlobHandle PopulateMarshallingBlob(MetadataBuilder builder)
+        {
+            var blobBuilder = new BlobBuilder();
+            SerializeMarshallingDescriptor(blobBuilder);
+            return builder.GetOrAddBlob(blobBuilder);
+
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs:199]:
```suggestion
```

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs (hunk 3 above)
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs:62]: Right

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs (hunk 4 above)
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs]: Moved

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs [hunk 6]
```diff
@@ -51,7 +54,22 @@ internal sealed class MethodBuilderImpl : MethodBuilder
 
         protected override bool InitLocalsCore { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
         protected override GenericTypeParameterBuilder[] DefineGenericParametersCore(params string[] names) => throw new NotImplementedException();
-        protected override ParameterBuilder DefineParameterCore(int position, ParameterAttributes attributes, string? strParamName) => throw new NotImplementedException();
+        protected override ParameterBuilder DefineParameterCore(int position, ParameterAttributes attributes, string? strParamName)
+        {
+            if (position > 0 && (_parameterTypes == null || position > _parameterTypes.Length))
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs:59]:
```suggestion
            if (_parameterTypes == null || position > _parameterTypes.Length)
```

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs [hunk 7]
```diff
@@ -51,7 +54,22 @@ internal sealed class MethodBuilderImpl : MethodBuilder
...
+                throw new ArgumentOutOfRangeException(SR.ArgumentOutOfRange_ParamSequence);
+
+            if (position == 0 && _parameters == null)
+            {
+                _parameters = new ParameterBuilderImpl[1];
+            }
+
+            Debug.Assert(_parameters != null);
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs:67]:
```suggestion
```

**Review by jkotas — APPROVED** [review]
Thanks

**Comment by buyaa-n** [comment]
The failures are unrelated and looks all timed out infra issue tracked with https://github.com/dotnet/runtime/issues/76454

**Pull request merged**